launch/test/launch/test_execute_local.py

```diff
@@ -126,6 +126,38 @@ def generate_launch_description():
     assert expected_called_count == on_exit_callback.called_count
 
 
+def test_execute_process_with_respawn_max_retries():
+    """Test launching a process with respaw_max_retries attribute."""
+    def on_exit_callback(event, context):
+        on_exit_callback.called_count = on_exit_callback.called_count + 1
+    on_exit_callback.called_count = 0
+
+    respawn_max_retries = 2   # we want the process to respawn this amount of times
+    shutdown_time = 10.0   # leave enough time to actually test the respawn_max_retries
```

Comment:
So the problem with these tests is that they will *always* take 10 seconds to expire, even when we were successful after a very short amount of time.  That's not great for the buildfarm.

Can we instead do something smarter in the `on_exit_callback`, and have it quit early if we get the expected number of calls before the timeout?

Reply:
Sure thing, I didn't do it that way at first because I didn't want to disrupt with the other tests.
Please take a look at the new approach and let me know what you think. @clalancette 

Basically what I did was trigger a Timer action (which would trigger a Shutdown signal) on the `on_exit_callback` when the `expected_called_count` is reached. I believe this is better than just trigger a Shutdown action because by waiting for a few seconds before shutting down we are really testing whether or not the `respawn_max_retries` quantity is being respected, i.e. the process doesn't keep respawning itself when it shouldn't.
I decided to leave the original 10secs timer as a safeguard in case something goes wrong one day and the `expected_called_count` is never reached.

Test results comparisson:
With this changes both the `test_execute_local`  and `test_execute_process` tooked 2secs, 5x less than before (see previous images [here](https://github.com/ros2/launch/pull/696#issuecomment-1691849061)).

![image](https://github.com/ros2/launch/assets/95138114/d3e11f8f-56b0-412d-aa81-adf4c4561145)

![image](https://github.com/ros2/launch/assets/95138114/54f7a70f-de12-4842-8b87-472e195e6b77)